### ISSUE TITLE
Parse a gatorgrader.yml file and run multiple GatorGrader Checks

### ISSUE THREAD

**Issue by gkapfham** [GatorEducator/gatorgrader#219]
**Is your feature request related to a problem? Please describe.**

For individuals who are not programming in Java, it is potentially confusing that they currently have to use GatorGrader's Gradle plugin to run all of the checks for an assignment. This has the potential for even more confusion if an individual is not programming in either Java or Python. Would it be possible for GatorGrader to read in a `gatorgrader.yml` file and then parse that file and run all of the pre-configured checks --- all through the same Python program? Also, would it then be useful if GatorGrader was available on PyPI so that you could run a command like `pip install gatorgrader` and then be able to install the program?

**Describe the solution you'd like**

- Make GatorGrader a program that is available on PyPI and supports installation with programs like `pip`
- Add features that would allow GatorGrader to parse a `gatorgrader.yml` configuration file
- Add features that would start multiple GatorGrader forked processes that can run each specified check
- Add features that would, if  possible, run each specified check in parallel
- Add features that would collect the result from each check and display it to the user

**Describe alternatives you've considered**

It is worth noting that it is a good idea to maintain and continue to use the Gradle plugin for assignments involved with the implementation and testing of a Java program.

**Technologies for consideration**

- YAML parsing with [PyYAML](https://pyyaml.org/)

**Comment by enpuyou** [comment]
Hi Professor @gkapfham, I was just looking into the `gatorgrader.yml` and its related parsing functions in java. And I have a few questions:

Here is a section of the `gatorgrader.yml` I got from a 101 lab. Is there any particular reason why the paths are structured like this for multiple files?
```
src/main:
    java:
        labseven/experiment/Experiment.java:
            --multi 2 --language Java
            --fragment "TODO" --count 0 --exact
            --fragment "println(" --count 8 --exact
src/main:
    java:
        labseven/experiment/RunCampaign.java:
            --multi 4 --language Java
            --fragment "TODO" --count 0 --exact
            --fragment "println(" --count 3 --exact
``` 
Can it be:
```
src/main/java/labseven/experiment:
      Experiment.java:
            --multi 2 --language Java
            --fragment "TODO" --count 0 --exact
            --fragment "println(" --count 8 --exact
      RunCampaign.java:
            --multi 4 --language Java
            --fragment "TODO" --count 0 --exact
            --fragment "println(" --count 3 --exact
``` 

In addition to that, I have been receiving some other parsing errors while using `PyYaml`, mostly due to structure like following where having scalars not mapped to any preceding keys or other scalars I'm guessing:
```
CountCommits --count 10
ExecuteCommand --command "gradle build"
ExecuteCommand --command "gradle -q --console plain run"
CountCommandOutput --command "gradle -q --console plain run" --count 4 --exact
```
would it be better to change to this as a sequence of scalars later on?
```
- CountCommits --count 10
- ExecuteCommand --command "gradle build"
- ExecuteCommand --command "gradle -q --console plain run"
- CountCommandOutput --command "gradle -q --console plain run" --count 4 --exact
```

Thank you very much!

**Comment by Michionlion** [comment]
Hi @enpuyou,
The GatorGrader.yml file is actually not strictly YAML, which is probably why some of the issues you have are cropping up; as you said, the "standalone" checks are just listed, each on their own line, which is not valid YAML.
As for paths, you can check out the [current implementation](https://github.com/GatorEducator/gatorgradle/blob/master/src/main/java/org/gatorgradle/config/GatorGradleConfig.java) to see how things are actually processed, but the gist of it is that any path setup, as you listed, is possible -- they are simply added together (with appropriate `/`s added) to get the final path that the ending lines.

**Comment by enpuyou** [comment]
> Hi @enpuyou,
> The GatorGrader.yml file is actually not strictly YAML, which is probably why some of the issues you have are cropping up; as you said, the "standalone" checks are just listed, each on their own line, which is not valid YAML.
> As for paths, you can check out the [current implementation](https://github.com/GatorEducator/gatorgradle/blob/master/src/main/java/org/gatorgradle/config/GatorGradleConfig.java) to see how things are actually processed, but the gist of it is that any path setup, as you listed, is possible -- they are simply added together (with appropriate `/`s added) to get the final path that the ending lines.

Thanks for the confirmation @Michionlion ! That's what I thought but I wasn't sure. I was thinking if we want to proceed with `PyYAML`, it would be better to make `gatorgrader.yml` to have the valid YAML syntax, that might also require some extra update to the Gradle side of the code.

**Comment by Michionlion** [comment]
The issue also would be backwards compatibility, which we should really try to maintain as much as possible (we for sure want old labs to work with any new changes).

**Comment by enpuyou** [comment]
@Michionlion That's right! Thanks for pointing that out. I think the python implementation could be just designed for the valid syntax YAML, while the Gradle code should be able to handle both. I haven't looked deep into the current Gradle implementation yet, but I think it's definitely feasible to have it compatible with the new and old structures.

**Comment by bagashvilit** [comment]
Hi everyone, we have been discussing using Poetry to answer one of the points in this issue: how to make GatorGrader a program that is available on PyPI and support installation with programs like `pip`. The reason why Poetry may be a better option is that it combines dependency management, packaging, and publishing into a single configuration file. Meaning that in addition to easily publishing the project to PyPI we may also avoid maintaining files like `setup.py`, `requirements.txt`, `setup.cfg`, `Pipfile` by combining all the metadata, dependencies, scripts, or entry points in Poetry `pyproject.toml`([Poetry documentation](https://github.com/python-poetry/poetry#introduction))

To migrate from current set up to Poetry set up we might want to use [dephell](https://dephell.readthedocs.io/cmd-deps-convert.html) or [pipenv-poetry-migrate](https://pypi.org/project/pipenv-poetry-migrate/). These tools will at least ensure that we convert dependencies from `Pipfile` to `pyproject.toml` format, but we will have to be careful that scripts are still executed properly. I will also be creating a separate branch for this.

**Comment by donizk** [comment]
Hi everyone! Like @kailaniwoodard said, after having met with both her and @bagashvilit , I chose to answer the question relating to the commands that are equivalent between pipenv and poetry. 

**What are the equivalent commands between pipenv and poetry?**
Both Poetry and Pipenv have the ability to install dependencies (as part of being dependency management tools), but there are some differences in terms of the commands used and how they function. I will break down the equivalent commands and note the differences in the usage of them (if there are any). 

These are just some of the equivalent commands that I was able to find through my research of both poetry and pipenv. 

**Installing Dependencies (Packages)**
`poetry install`: this command will install all dependencies (including development dependencies)
`pipenv install`: this command will install all non-dev dependencies

*Note: to only install the normal dependencies in poetry, you can use `poetry install --no-dev`

**Adding Non-Dev Dependencies**
- these commands will install a non-dev dependency that is specified where it says packageName.
`poetry add packageName`
`pipenv install packageName`

**Adding Dev Dependencies**
- these commands will install a dev dependency
`poetry add --dev packageName`
`pipenv install --dev packageName`: this command will install a dev dependency

**Locking Dependencies**
- these commands will lock a file
`poetry lock`
`pipenv lock`

**Uninstalling Dependencies**
- these commands will remove a dependency 
`poetry remove packageName`
`pipenv uninstall packageName`

*Note: to uninstall a dev dependency with poetry, you must use this command: `poetry remove --dev packageName`

**Checking versions**
- these commands will check the version of the tool being used
`poetry --version`
`python --version`

**Activating shell**
- these commands will "spawn" a shell within the virtual environment
`poetry shell`
`pipenv shell`

**Comment by corlettim** [comment]
@bagashvilit @kailaniwoodard @donizk, you all did phenomenal research this past week!! All of this information will not only be beneficial in the long run when it comes time for implementation but also for the other developers that may not be as familiar with Poetry. Keep up the great work!!

**Comment by gkapfham** [comment]
Hi everyone! What is the consensus? Do we think that it would be a good idea to implement this feature? If you think that this
feature is useful, can you comment briefly on why you think it is most useful? Does anyone have a list of steps that we would need to follow in order to complete this features?

**Comment by Michionlion** [comment]
Hi all! This issue has been resolved with the implementation of [GatorGrade](https://github.com/GatorEducator/gatorgrade), and PyPI package publishing will be completed before the start of this upcoming semester.
